ENH: Add actions to fetch data from BV-BRC.

VinzentRisch opened the pull request:
closes #200 

- Adds new action get_bv_brc_genomes to fetch genome sequences outputted as GenomeData[DNASequences] with corresponding taxonomy data as FeatureData[Taxonomy].
- Adds new action get_bv_brc_metadata to fetch any type of metadata outputted as immutable metadata.
- Adds new action get_bv_brc_genome_features to fetch genome feature sequences outputted as GenomeData[Genes] and GenomeData[Proteins]. Additionally taxonomy as FeatureData[Taxonomy] is provided and GenomeData[Loci]

- All actions make use of the [Data API](https://www.bv-brc.org/api/doc/) by BV-BRC. 
- The data that should be downloaded can be queried with three options. First for all actions an RQL style query can be provided. Second by providing IDs/values and a corresponding data field, you can retrieve all data associated with those specific values in that data field. And as a third option a metadata column can be provided, to use metadata obtained with the action get-bv-brc-metadata as a new query. 

### Run it locally 
1. First, clone the repo and checkout the PR branch:
```bash
git clone https://github.com/bokulich-lab/rescript.git
cd RESCRIPt
git fetch origin pull/202/head:pr-202
git checkout pr-202
pip install -e .
```

3. Test it out!
```bash
qiime rescript get-bv-brc-metadata --p-data-type genome --p-rql-query "in(genome_id,(224308.43,2030927.4755))" --output-dir metadata_rql
qiime rescript get-bv-brc-metadata --p-data-type genome --p-ids 224308.43 2030927.4755 --p-data-field "genome_id" --output-dir metadata_ids 
qiime rescript get-bv-brc-metadata --p-data-type genome --m-ids-metadata-file  metadata_ids/metadata.qza  --m-ids-metadata-column "species" --output-dir metadata_mcolumn

qiime rescript get-bv-brc-genomes --p-rql-query "in(genome_id,(224308.43,2030927.4755))" --output-dir genome_sequences_rql
qiime rescript get-bv-brc-genomes --p-ids 224308.43 2030927.4755 --p-data-field "genome_id" --output-dir genome_sequences_ids
qiime rescript get-bv-brc-genomes --m-ids-metadata-file  metadata_ids/metadata.qza  --m-ids-metadata-column genome_id --output-dir genomes_mcolumn

qiime rescript get-bv-brc-genome-features --p-rql-query "in(genome_id,(2030927.4755))" --output-dir features_rql
qiime rescript get-bv-brc-genome-features --p-ids 224308.43 2030927.4755 --p-data-field "genome_id" --output-dir features_ids
qiime rescript get-bv-brc-genome-features --m-ids-metadata-file  metadata_ids/metadata.qza  --m-ids-metadata-column genome_id --output-dir features_mcolumn


```

mikerobeson commented:
When I try and install this branch I obtain the following error:

> from q2_types.genome_data import GenomeData, Loci, Proteins, Genes, DNASequence
ImportError: cannot import name 'DNASequence' from 'q2_types.genome_data' (/home/.../.conda/envs/qiime2-amplicon-2024.5-test/lib/python3.9/site-packages/q2_types/genome_data/__init__.py)

I suggest that you merge this `get_bv_brc_data` branch with the latest version of the master / main branch the RESCRIPt repo. This PR is currently **3 commits behind the current branch**, and fails with the above error.

VinzentRisch commented:
Hi @mikerobeson 
I updated the branch. Now it should be ready to review.

mikerobeson commented:
I think there is still an issue. I am still unable to install the plugin and run `qiime --help`. When I do I get:

>from q2_types.genome_data import GenomeData, Loci, Proteins, Genes, DNASequence
ImportError: cannot import name 'DNASequence' from 'q2_types.genome_data' (/Users/.../AnalysisSoftware/miniconda3/envs/qiime2-amplicon-2024.5-test/lib/python3.9/site-packages/q2_types/genome_data/__init__.py)

mikerobeson commented on rescript/plugin_setup.py:
```diff
@@ -1235,6 +1238,149 @@
+    output_descriptions={
+        'metadata': 'metadata'},
+    name='Fetch BV-BCR metadata.',
+    description="Fetch BV-BCR metadata for a specific data type with an RQL "
```
Again, please provide a detailed description of what this action retrieves and what it outputs. What is the downstream use case for this? Can you provide an example?

mikerobeson commented on rescript/plugin_setup.py:
```diff
@@ -1235,6 +1238,149 @@
+    outputs=[('genomes', GenomeData[DNASequence])],
+    input_descriptions={},
+    parameter_descriptions={
+        'rql_query': 'Query in RQL format. Check '
+                     'https://www.bv-brc.org/api/doc/genome_sequence '
+                     'for documentation.',
+        'genome_ids': 'List of genome IDs from BV-BRC.',
```
Define and describe `rql_query` and `genome_ids`. For example, are the latter GenBank accessions?

mikerobeson commented on rescript/plugin_setup.py:
```diff
@@ -1235,6 +1238,149 @@
+
+    },
+    name='Fetch taxonomy data from BV-BRC.',
+    description='Fetch taxonomy data from BV-BRC.',
```
Again, provide detailed description.

mikerobeson reviewed (changes requested):
Great start @VinzentRisch! I was able to run your example code via the dev version of QIIME 2.

There should be more content within the `parameter_descriptions` and `descriptions`. Also, it would be nice to have as part of your PR description, and the code help text / descriptions, the "use cases" for these actions. That is, what does a user do once they have these QZA files? How do they work with it? Perhaps consider constructing a RESCRIPt tutorial to eventually go along with this?

I'd also recommend swapping the names of your actions so that the name for the reference data is first: 

  - bv-brc-fetch-genome-features
  - bv-brc-fetch-genomes
  - bv-brc-fetch-metadata
  - bv-brc-fetch-taxonomy

mikerobeson commented on rescript/plugin_setup.py:
```diff
@@ -1235,6 +1238,149 @@
+    output_descriptions={
+        'genomes': 'genomes',
+    },
+    name='fetch genomes',
```
Should be consistent with the other names, and probably be something like `Fetch Genomes from BV-BRC`.

VinzentRisch commented:
Hi @mikerobeson 
- So I now removed the taxonomy action and added taxonomy outputs to the actions for genomes and genome features. I think having three actions makes it easy enough but still flexible to be specific with data querying.
- I added more options to query the data more effectively. 
- I have not yet added usage examples to the descriptions. I will come up with some examples on Monday during our dev meeting. Because I am not so sure myself what all can be done with this data in Qiime.

mikerobeson commented:
Hi @VinzentRisch,  I apologize for taking a while to get to this. 😬 

I really like to improvements. I was able to run everything w/o issue. Again, as I've never really used this tool, I'm not sure what the next steps are with the QZA outputs. Once I have an idea of what the "final" steps are I should be able to finalize my review. Other than that, nothing really stands out to me to fix.. at least for the moment. :-)

mikerobeson commented:
Thanks for that @nbokulich! I am okay to merge this as it is, if you think it is ready. I've nothing more to add. 👍
